pkg/controller/nodeipam/ipam/cloud_cidr_allocator.go

```diff
@@ -311,20 +343,48 @@ func (ca *cloudCIDRAllocator) updateCIDRAllocation(nodeName string) error {
 
 	cidrStrings := make([]string, 0)
 
-	if len(instance.NetworkInterfaces) == 0 || (len(instance.NetworkInterfaces) == 1 && len(instance.NetworkInterfaces[0].AliasIpRanges) == 0) {
+	// No Pod CIDRs can be allocated to a node, if there are no network interfaces;
+	// or if there is one network interface, but it has neither an IP alias nor an IPv6 address.
+	if len(instance.NetworkInterfaces) == 0 ||
+		(len(instance.NetworkInterfaces) == 1 &&
+			len(instance.NetworkInterfaces[0].AliasIpRanges) == 0 &&
+			ca.cloud.GetIPV6Address(instance.NetworkInterfaces[0]) == nil) {
+
 		nodeutil.RecordNodeStatusChange(ca.recorder, node, "CIDRNotAvailable")
 		return fmt.Errorf("failed to allocate cidr: Node %v has no ranges from which CIDRs can be allocated", node.Name)
 	}
 
-	// sets the v1.NodeNetworkUnavailable condition to False
+	// Sets the v1.NodeNetworkUnavailable condition to False.
 	ca.setNetworkCondition(node)
 
-	// nodes in clusters WITHOUT multi-networking are expected to have only 1 network-interface with 1 alias IP range.
-	if len(instance.NetworkInterfaces) == 1 && len(instance.NetworkInterfaces[0].AliasIpRanges) == 1 {
-		cidrStrings = append(cidrStrings, instance.NetworkInterfaces[0].AliasIpRanges[0].IpCidrRange)
-		ipv6Addr := ca.cloud.GetIPV6Address(instance.NetworkInterfaces[0])
-		if ipv6Addr != nil {
-			cidrStrings = append(cidrStrings, ipv6Addr.String())
+	// Nodes in clusters WITHOUT multi-networking are expected to have only 1 network-interface
+	// with 1 alias IPv4 range and/or 1 IPv6 address. Multi-network clusters may have 1 interface
+	// with multiple aliases.
+	if len(instance.NetworkInterfaces) == 1 &&
+		(len(instance.NetworkInterfaces[0].AliasIpRanges) == 1 ||
+			ca.cloud.GetIPV6Address(instance.NetworkInterfaces[0]) != nil) {
+
+		ipv4CIDR := ""
+		if len(instance.NetworkInterfaces[0].AliasIpRanges) > 0 {
+			ipv4CIDR = instance.NetworkInterfaces[0].AliasIpRanges[0].IpCidrRange
+		}
+
+		ipv6CIDR := ""
+		if addr := ca.cloud.GetIPV6Address(instance.NetworkInterfaces[0]); addr != nil {
+			ipv6CIDR = addr.String()
+		}
+
+		switch {
+		case ca.stackType == stackIPv4 && ipv4CIDR != "":
+			cidrStrings = []string{ipv4CIDR}
+		case ca.stackType == stackIPv4IPv6 && ipv4CIDR != "" && ipv6CIDR != "":
+			cidrStrings = []string{ipv4CIDR, ipv6CIDR}
+		case ca.stackType == stackIPv6IPv4 && ipv4CIDR != "" && ipv6CIDR != "":
+			cidrStrings = []string{ipv6CIDR, ipv4CIDR}
+		case ca.stackType == stackIPv6 && ipv6CIDR != "":
+			cidrStrings = []string{ipv6CIDR}
+		default:
+			return fmt.Errorf("failed to allocate cidr: Node %v has no ranges from which CIDRs can be allocated for the cluster stack family %s", node.Name, ca.stackType)
 		}
 	} else {
 		// multi-networking enabled clusters
```

Comment:
I think you need to handle the multi networking case as well. Otherwise we may end up assigning an IPv6 address even if the cluster is not dual stack.

Reply:
Could you please elaborate on why you think the current approach won't work correctly?

The condition
```
	if len(instance.NetworkInterfaces) == 1 &&
		(len(instance.NetworkInterfaces[0].AliasIpRanges) == 1 ||
			ca.cloud.GetIPV6Address(instance.NetworkInterfaces[0]) != nil) {
```

evaluates as true only for a single network interface that has one alias ip range or an ipv6 address.

If there are multiple network interfaces, this code won't execute. If there is a single network interface with multiple alias ip ranges (what is possible for multi-network as well), the code won't execute either. I assume there is yet another case related to the ipv6 addresses? If a single multi-network interface can have multiple alias ip ranges, maybe it could also have multiple ipv6 addresses assigned? Is this what you meant?

Reply:
My point was that we need to fix this for multi subnet clusters as well. i.e. the case where a single stack cluster is running on a dual stack subnet and has multiple interfaces. We will end up populating pod CIDR for IPv6 as well in that case since it runs the old logic.

Reply:
@sdmodi IIRC the logic in ultinetwork uses a different source of truth to obtain the cidrs

> cidrStrings, err = ca.performMultiNetworkCIDRAllocation(node, instance.NetworkInterfaces, hasNodeLabels)

Reply:
I haven't checked the code for a while, but I recall that it will find the right IPv4 alias IP over there. But it still won't honor stack type before assigning an IPv6 podCIDR to the node. For single stack IPv6 clusters, it may still work because you will not see a default pod range label on the node. It still fails for the case where a single stack IPv4 cluster is on a dual stack subnet.

Reply:
@sdmodi is multinetwork compatible with dual-stack clusters? I wonder because of https://sourcegraph.com/github.com/kubernetes/cloud-provider-gcp@4e278fc87322ab7a930de3556fc5b8b752e9295d/-/blob/pkg/controller/nodeipam/ipam/multinetwork_cloud_cidr_allocator.go?L76

While I don't really understand the whole code here yet, it seems to me two places mimic the behavior I'm changing for single-network. These are: https://sourcegraph.com/github.com/kubernetes/cloud-provider-gcp@4e278fc87322ab7a930de3556fc5b8b752e9295d/-/blob/pkg/controller/nodeipam/ipam/multinetwork_cloud_cidr_allocator.go?L102 and https://sourcegraph.com/github.com/kubernetes/cloud-provider-gcp@4e278fc87322ab7a930de3556fc5b8b752e9295d/-/blob/pkg/controller/nodeipam/ipam/multinetwork_cloud_cidr_allocator.go?L151 So maybe they need a similar change?

Reply:
Yes. That is exactly what I was calling out. Both places need to check for stack type.

Reply:
@sdmodi I made the changes that you requested, however unfortunately there are no unit tests for IPv6 addresses with the multi-network code. This makes it really hard for me to make changes to the code, because I am not familiar with the code base. Can you point me to any documentation for multi-network so I could understand how it works? I wonder that maybe it is better to only make changes to the single-network code for now and leave multi-network for a follow-up PR, once the missing tests are added? Please let me know your thoughts.

Reply:
Please work with @yvetteli0314 on this. It is fine to make the changes for single network code for now and multi network support later.

Reply:
Ok, I'll reach out to @yvetteli0314 after we merge this PR.